CodingGuidelines.md

```diff
@@ -0,0 +1,507 @@
+    /// AudiorFormat lists the supported / known formats of audio data
+    /// </summary>
+    public enum AudioFormat
+    {
+        /// <summary>
+        /// No specified format
+        /// </summary>
+        None = 0,
+        /// <summary>
+        /// Undefined format.
+        /// </summary>
+        Other,
+        /// <summary>
+        /// PCM
+        /// </summary>
+        PCM,
+        /// <summary>
+        /// MP3.
+        /// </summary>
+        MP3,
+        /// <summary>
+        /// Ogg Vorbis
+        /// </summary>
+        Ogg
+    }
+```
+
+## Review Enum use for Bitfields.
+
+If there is a possibility for an enum to require multiple states as a value, e.g. Handedness = Left & Right. Then the Enum needs to be decorated correctly with BitFlags to enable it to be used correctly
+
+> The Handedness.cs file has a concrete implementation for this
+
+### Don't:
+
+```
+public enum MyEnum
+{
+    None,
+    Left,
+    Right
+}
+```
+
+### Do:
+
+ ```
+ [flags]
+public enum MyEnum
+{
+    None = 1  << 0,
+    Left = 1  << 1,
+    Right = 1 << 2
+}
+ ```
+
+
+## Best Practices, including Unity recommendations
+
+Some of the target platforms of this project require us to take performance into consideration.  With this in mind we should always be careful of allocating memory in frequently called code in tight update loops or algorithms.
+
+## Encapsulation
+
+Always use private fields and public properties if access to the field is needed from outside the class or struct.  Be sure to co-locate the private field and the public property. This makes it easier to see, at a glance, what backs the property and that the field is modifiable by script.
+
+If you need to have the ability to edit your field in the inspector, it's best practice to follow the rules for Encapsulation and serialize your backing field.
+
+>The only exception to this is for data structures that require the fields to be serialized by the `JsonUtility`, where a data class is required to have all public fields for the serialization to work.
+
+### Don't:
+
+```
+public float MyValue;
+```
+
+### Do:
+
+ ```
+ // private field, only accessible within script (field is not serialized in Unity)
+ private float myValue;
+  ```
+
+### Do:
+
+ ```
+ // Enable private field to be configurable only in editor (field is correctly serialized in Unity)
+ [SerializeField] 
+ private float myValue;
+  ```
+
+---
+
+ ### Don't:
+
+ ```
+ private float myValue1;
+ private float myValue2;
+ 
+ public float MyValue1
+ {
+     get{ return myValue1; }
+     set{ myValue1 = value }
+ }
+ 
+ public float MyValue2
+ {
+     get{ return myValue2; }
+     set{ myValue2 = value }
+ }
+```
+
+ ### Do:
+
+ ```
+ // Enable field to be configurable in the editor and available externally to other scripts (field is correctly serialized in Unity)
+ [SerializeField] 
+ private float myValue; // <- Notice we co-located the backing field above our corrisponding property.
+ public float MyValue
```

Comment:
missing summary tags in example. Ideally the public properties summary would match the tooltip of the serialized field.